### PR TITLE
[locale] ja, zh-*, ko: Standardize CJK time formats

### PR DESCRIPTION
**Pull request by jasonkb**
Time formats like '16:00' are predominant on China, Taiwan, Hong Kong, and Japan websites and apps. Time format like '오후 4:00' is predominant on Korea websites and apps. The current moment.js formats, following pattern of "4[시点點時] 0[分분]", are infrequently used for displaying times on CJK websites and apps. In particular, a time with zero minutes like "4시 0분", "4点0分", or "4時0分" is exceedingly awkward.

This change also removes an inconsistency in 'calendar' between zh-cn and the two other zh locales (zh-hk and zh-tw). Now 'calendar' will refer to days in the current week in the same way across these three locales. I am roughly familiar with written Chinese as used in CN, TW, and HK, and am not aware of a systematic difference how one refers to calendar days in the current week. Let me know if I'm missing something :)

This change also makes the l, ll, lll, llll formats set explicitly to match the L, LL, LLL, LLLL in CJK. This is to avoid incorrect day-of-week abbreviations like "2010年2月14日午後3時25分 日" ("日" is inappropriate as a standalone word) or "2010년 2월 14일 일 오후 3시 25분" ("일" is inappropriate as a standalone word).

@baryon (ja)
@kyungw00k @jeeeyul (ko)
@suupic @zenozeng (zh-cn)
@ben-lin @hehachris (zh-tw and zh-hk)
Could you help review? Thank you :)

Below I have collected some evidence that the formats in this PR are appropriate, from various popular websites and apps.

### ja

Facebook uses "16:00" format.
screenshot 2016-11-15 01 32 42

GMail uses "16:00" format.
screenshot 2016-11-15 23 20 38
screenshot 2016-11-15 23 20 44

Wikipedia talk pages use "16:00" format.
screenshot 2016-11-15 22 54 10

Line on Japanese-language phone uses "16:00" format.
15129010_10101031834192808_8544315342583548960_o

KakaoTalk on Japanese-language phone uses "16:00" format.
15025430_10101031836458268_223990604552638952_o

### ko

Facebook uses "오후 4:00" format.
screenshot 2016-11-15 01 49 35

Twitter uses "오후 4:00" format.
screenshot 2016-11-15 23 30 39

GMail uses "오후 4:00" in some places and "16:00" format in other places (!)
screenshot 2016-11-15 23 17 53
screenshot 2016-11-15 23 18 36

Naver News uses "16:00" format.
screenshot 2016-11-15 23 12 29

Line on Korean-language phone uses "16:00" format.
15110938_10101031839746678_5586062181360908786_o

KakaoTalk on Korean-language phone uses "16:00" format.
15068957_10101031840110948_6324736872530080588_o

Wikipedia talk pages use "16:00" format.
screenshot 2016-11-15 23 22 40

### zh-cn

Chinese-language websites (in Simplified and in Traditional) seem to be split about whether they use 12 or 24 hour formats. So it's a judgment call about which to use in moment.js. I recommend using 24 hour format in moment for succinctness.

People's Daily uses 24-hr format.
screenshot 2016-11-15 02 01 27

Taobao tracking status uses 24-hr format.
screenshot 2016-11-15 17 02 15

百度外卖 (Baidu delivery) uses 24-hr format.
default

Wikipedia talk pages use 24-hr format.
screenshot 2016-11-15 22 51 13

GMail uses 24-hr format some places and 12-hr format other places (!)
screenshot 2016-11-15 23 24 17
screenshot 2016-11-15 23 24 51

Facebook uses 12-hour format.
screenshot 2016-11-15 01 52 19

Twitter uses 12-hour format.
screenshot 2016-11-15 23 32 05

Alipay uses 12-hour format.
slack for ios upload

### zh-tw and zh-hk

Facebook uses "16:00" format.
screenshot 2016-11-15 01 52 50

Pixnet.net (seems to be a popular TW site) uses "16:00" format, at least in this promotion and blog timestamps.
screenshot 2016-11-15 23 37 07
screenshot 2016-11-15 23 38 03

hkgolden.com (seems to be a popular HK site) uses "16:00" format, at least in its forum timestamps and footer.
screenshot 2016-11-15 23 39 06
screenshot 2016-11-15 23 39 55

GMail uses 24-hr format some places and 12-hr format other places.
screenshot 2016-11-15 23 26 06
screenshot 2016-11-15 23 26 14

Twitter uses 12-hr format.
screenshot 2016-11-15 23 32 05

### CODE REVIEW

**Comment by sibnerian** [comment]
LGTM!

**Comment by spikebrehm** [comment]
This is amazing!  Thanks for being so thorough @jasonkb!

**Comment by ibread** [comment]
LGTM for 24hr format for `zh-cn`. Thx a lot for making this happen, @jasonkb !!

**Comment by liangg** [comment]
LGTM it actually looks amazing

**Comment by lencioni** [comment]
LGTM!

File: src/locale/zh-cn.js [hunk 1]
```diff
@@ -12,16 +12,16 @@ export default moment.defineLocale('zh-cn', {
...
-        LTS : 'Ah点m分s秒',
-        L : 'YYYY-MM-DD',
+        LT : 'HH:mm',
+        LTS : 'HH:mm:ss',
+        L : 'YYYY年MMMD日',
         LL : 'YYYY年MMMD日',
-        LLL : 'YYYY年MMMD日Ah点mm分',
-        LLLL : 'YYYY年MMMD日ddddAh点mm分',
-        l : 'YYYY-MM-DD',
+        LLL : 'YYYY年MMMD日 HH:mm',
```
Comment (maggiepint) [inline src/locale/zh-cn.js]: I sit next to a Chinese developer, and I asked him to look at this one. He feels that the first changes (LT, LTS, L) make sense, but the second changes (LLL, LLLL) don't. This makes sense to me as if you truly want a long format, it makes more sense to use a written-word style throughout. Thoughts?

Comment (jasonkb) [inline src/locale/zh-cn.js]: @maggiepint Thank you for the feedback!

The screenshots (in PR description) show that People's Daily, 百度外卖 (Baidu delivery), Wikipedia, GMail, Facebook, Twitter, and Alipay show that these popular apps and websites all prefer a 'YYYY年MMMD日 HH:mm' format (or a similar  format in 12-hour time). Maybe show those screenshots to your deskmate? :)  Thank you again.

Comment (baryon) [inline src/locale/zh-cn.js]: I agree maggiepint. 
In these sites/apps, you can't found `YYYY年MMMD日Ah点mm分`, just because these developer can't found a library like `moment` to format datetime easily.

Comment (jasonkb) [inline src/locale/zh-cn.js]: Sorry to try pushing back one more time, but I am especially concerned about the YYYY年MMMD日Ah点mm分 format when there are 0 minutes -- the relatively common case of a timestamp that is on the hour. This comes out as e.g. "2016年11月17日下午4点0分" which is quite awkward -- one would not write it that way, and certainly not speak it that way (the 0分 would be elided). That said, I'd love for this to be a data-driven decision, so if you have some evidence that YYYY年MMMD日Ah点mm分 is the preferred format on the internet, I'm happy to switch to the preferred format. Thank you again!

P.S. my deskmate :)  @ibread offers two more evidences that HH:MM is standard in longform timestamps, from China's top two email providers.
screenshot 2016-11-18 14 33 25
screenshot 2016-11-18 14 31 36

Comment (zenozeng) [inline src/locale/zh-cn.js]: @jasonkb, Yes `HH:mm` is more common. However, I agree with @maggiepint, `LLL` and `LLLL` should keep the long format version.  

#### LLL and LLLL

screenshot from 2016-11-18 16-08-03

In zh-CN, both `HH时mm分` and `HH:mm` is used, removing the original `LLL` and `LLLL`  might make it inconvenient for some cases. 

#### 0分

screenshot from 2016-11-18 16-24-45

As for the `0分` issue, we can just omit the `分` part in output when `this.minutes()  === 0`.

```
04: 00 => 4时
15: 40 => 15时40分
14: 12：36 => 14时12分36秒
```

#### See Also

- [GB/T 15835-2011](https://github.com/moment/moment/files/599331/GB-T-15835-2011.pdf) (China's General rules for writing numerals in publications)
- http://www.china-language.gov.cn/wenziguifan2/shanghi/021.htm

Comment (ibread) [inline src/locale/zh-cn.js]: @zenozeng Thx for following up! Would you mind giving a few examples that the new format of `LLL` and `LLLL` would cause inconvenience? 

If I understand correctly, the standard ([GB/T 15835-2011](https://github.com/moment/moment/files/599331/GB-T-15835-2011.pdf)) is for `when to use Chinese numerals vs. Arabic numbers` (一/二/三 vs. 1/2/3) in publications, not for `how to format time` in publications. This is shown in the preface of the standard, screenshot as below.

> This standard defines how to use Chinese numerals and Arabic numerals in publications.

screenshot 2016-11-18 17 46 02

To put it another way, I think the point of the standard is, when writing time like`4:12`, to enforce using `Arabic numerals` (4 and 12) instead of Chinese numbers (四 and 十二) in publication, not to enforce using `4时12分` over `4:12`.

Moreover, it actually suggests to use `GB/T7408-94 5.3.1.1 extended format when necessary`,  which users 24 hour format, like `04:00`, and is exactly what @jasonkb proposed here.

Comment (zenozeng) [inline src/locale/zh-cn.js]: @ibread Sorry, I wasn't too clear about what I had meant. Let me try clear bit.

There are a few tokens that can be used to format a moment based on its locale:
- LT
- LTS
- L
- LL
- LLL
- l
- ll
- lll
- llll

The lowercase version is intended to be the shortened version of its uppercase counterpart.

screenshot from 2016-11-18 19-04-00


I am not against using `HH:mm` for `lll` and `llll`, that looks good to me.
But as for `LLL` and `LLLL`, I think may be using `HH时mm分` is better. 

For Example:

```
Date: 2006-01-02 15:04:05
lll: 2006年1月2日 15:04
LLL: 2006年1月2日 15时4分 
```

```
Date: 2006-01-02 16:00:00
lll: 2006年1月2日 16:00
LLL: 2006年1月2日 16时
```

I am referencing `GB/T 15835-2011` because it provides some cases of writing DateTime in Chinese. 
I did not mean we _should_ use format like `2006年1月2日 15时4分`, I mean we _could_ use that format and it is not uncommon.

Like Microsoft Excel, we should provide both options and let user choose what they want.  

screenshot from 2016-11-18 19-09-38

Comment (spikebrehm) [inline src/locale/zh-cn.js]: Perhaps it's possible to define a format as as function, which takes the moment instance as an argument?

However that doesn't seem to be the case, looking at
https://github.com/moment/moment/blob/497f918515ae6ab900f008f19523b1e4ae5e2450/src/lib/locale/formats.js#L10

**Comment by patyoon** [comment]
LGTM for Korean

File: src/locale/zh-cn.js (hunk 1 above)
Comment (zenozeng) [inline src/locale/zh-cn.js]: @jasonkb Unfortunately, it seems not possible without modifying  `expandFormat` and `formatMoment`.  localFormattingTokens are just strings now. Maybe create another pull request to allow `LLL` and `LLLL` to be functions?

See: https://github.com/moment/moment/blob/eb337b9df0a844c4cfdfa5f7e08e58bd91b8874f/src/lib/format/format.js#L64-L91

Comment (zenozeng) [inline src/locale/zh-cn.js]: @jasonkb I will create a PR for localFormattingTokens later today.

Comment (zenozeng) [inline src/locale/zh-cn.js]: @jasonkb See #3612

Comment (ibread) [inline src/locale/zh-cn.js]: @zenozeng I see, this makes sense. Thx for the explanation. :)

**Comment by kentobento** [comment]
japanese lgtm.

File: src/locale/zh-cn.js (hunk 1 above)
Comment (jasonkb) [inline src/locale/zh-cn.js]: @zenozeng How do you want to order our two PRs? Should yours be merged first, then mine on top of yours (after I resolve conflicts)?

Comment (zenozeng) [inline src/locale/zh-cn.js]: @jasonkb Maybe yours first? The changes to `LT, LTS, L, l, lll, llll` looks good to me. As for `LLL` and `LLLL`, just keeping the original version is enough.

As for omitting `分` in `LLL` and `LLLL`, what about another pull request after #3612 merged?

**Comment by kyungw00k** [comment]
@jasonkb LGTM!

File: src/locale/zh-cn.js [hunk 2]
```diff
@@ -12,16 +12,16 @@ export default moment.defineLocale('zh-cn', {
...
     longDateFormat : {
-        LT : 'Ah点mm分',
-        LTS : 'Ah点m分s秒',
-        L : 'YYYY-MM-DD',
+        LT : 'HH:mm',
+        LTS : 'HH:mm:ss',
+        L : 'YYYY年MMMD日',
         LL : 'YYYY年MMMD日',
         LLL : 'YYYY年MMMD日Ah点mm分',
         LLLL : 'YYYY年MMMD日ddddAh点mm分',
```
Comment (jasonkb) [inline src/locale/zh-cn.js]: @zenozeng @baryon @maggiepint @ibread @spikebrehm Thank you again for the great feedback and help. Based on our discussion:

1) I've made this PR leave LLL and LLLL in zh-cn as it was.
2) @zenozeng will follow up with https://github.com/moment/moment/pull/3612 on top of this change to remove the "0分" from LLL and LLLL.

I'm really excited about this. Thank you again! @maggiepint what is the next step to merging this?

**Comment by jasonkb** [comment]
@maggiepint Is it possible to merge this PR now? Thank you~

**Comment by jasonkb** [comment]
@ichernev What do you recommend for next steps to merge this branch? Thank you!

**Comment by jasonkb** [comment]
@maggiepint @ichernev circling back in here, is this PR ready to merge? Thank you!

**Comment by sibnerian** [comment]
Bump! It would be fantastic to have these improved translations. @maggiepint @ichernev @mj1856

**Comment by maggiepint** [comment]
Okay, I haven't picked through every line of this, but judging by the sign offs, and given none of us speak any of these languages, will greenlight.

**Comment by jasonkb** [comment]
Thank you @maggiepint !

**Comment by ichernev** [comment]
This will get merged in next release, thank you @jasonkb

**Comment by jasonkb** [comment]
Thank you @ichernev!!!

**Comment by jasonkb** [comment]
@ichernev Thanks again @ichernev. Any idea of timeline for next release? :)

**Comment by ichernev** [comment]
Merged in 2d5c9f6112afefe0538064b4d49c32b71531002f
